csharp/src/Google.Protobuf/CodedInputReader.cs

```diff
@@ -0,0 +1,879 @@
+        public int ReadSFixed32()
+        {
+            return (int)ReadRawLittleEndian32();
+        }
+
+        /// <summary>
+        /// Reads an sfixed64 field value from the stream.
+        /// </summary>   
+        public long ReadSFixed64()
+        {
+            return (long)ReadRawLittleEndian64();
+        }
+
+        /// <summary>
+        /// Reads an sint32 field value from the stream.
+        /// </summary>   
+        public int ReadSInt32()
+        {
+            return CodedInputStream.DecodeZigZag32(ReadRawVarint32());
+        }
+
+        /// <summary>
+        /// Reads an sint64 field value from the stream.
+        /// </summary>   
+        public long ReadSInt64()
+        {
+            return CodedInputStream.DecodeZigZag64(ReadRawVarint64());
+        }
+
+        /// <summary>
+        /// Reads a length for length-delimited data.
+        /// </summary>
+        /// <remarks>
+        /// This is internally just reading a varint, but this method exists
+        /// to make the calling code clearer.
+        /// </remarks>
+        public int ReadLength()
+        {
+            return (int)ReadRawVarint32();
+        }
+
+        /// <summary>
+        /// Peeks at the next tag in the stream. If it matches <paramref name="tag"/>,
+        /// the tag is consumed and the method returns <c>true</c>; otherwise, the
+        /// stream is left in the original position and the method returns <c>false</c>.
+        /// </summary>
+        public bool MaybeConsumeTag(uint tag)
+        {
+            uint previousTag = lastTag;
+            long consumed = reader.Consumed;
+
+            uint newTag = ReadTag();
+            if (newTag == tag)
+            {
+                // Match so consume tag
+                return true;
+            }
+
+            // No match so rewind
+            long rewindCount = reader.Consumed - consumed;
+            if (rewindCount > 0)
+            {
+                reader.Rewind(rewindCount);
+                lastTag = previousTag;
+            }
+
+            return false;
+        }
+
+        #endregion
+
+        #region Underlying reading primitives
+
+        /// <summary>
+        /// Reads a raw Varint from the stream.  If larger than 32 bits, discard the upper bits.
+        /// This method is optimised for the case where we've got lots of data in the buffer.
+        /// That means we can check the size just once, then just read directly from the buffer
+        /// without constant rechecking of the buffer length.
+        /// </summary>
+        internal uint ReadRawVarint32()
+        {
+            byte value;
+
+            ThrowEndOfInputUnless(reader.TryRead(out value));
+            int tmp = value;
+            if (tmp < 128)
+            {
+                return (uint)tmp;
+            }
+            int result = tmp & 0x7f;
+            ThrowEndOfInputUnless(reader.TryRead(out value));
+            tmp = value;
+            if (tmp < 128)
```

Comment:
Why is this implemented as a nested if/else whereas ReadRawVarint64 is implemented as a loop? Is there a performance difference?

Reply:
I assume so. This mirrors the input stream logic.